Map mesh quantity to tracer particles in output HDF5 snapshot

hfhsieh opened the pull request:
Usage:
- Set `OPT__OUTPUT_PAR_MESH = 1` in `Input__Parameter`
- Add the labels of the fields to be outputted in the file `Input__Par_Mesh`
	- An example of `Input__Par_Mesh` is located in `example/input`

Done:
- Add four data members `Mesh_Attr_*` in the `Particle` structure
- Add a runtime parameter `OPT__OUTPUT_PAR_MESH` controlling whether to dump additional attributes mapped from mesh for tracer particles
- The target fields are specified in the file `Input__Par_Mesh` at runtime
	- Supports the active, passive, and derived fields available in `Prepare_PatchData()`
- Only support HDF5 snapshots
	- The data are in the `Particle` group with name in the format of `Mesh[FieldLabel]`
	- For non-tracer particles, the value is set to `__FLT_MAX__`
- Update `Makefile_base` to include the new file `Particle/Par_Output_TracerParticle_Mesh.cpp`

Future works:
- Update the doc/wiki
- Support derived fields defined `Flu_DerivedField_BuiltIn.cpp` and `Flu_DerivedField_User.cpp`
- (Optional) Support separate interpolation schemes used in `Par_UpdateTracerParticle()` and `Par_Output_TracerParticle_Mesh()`

Related Issues: #334 #361

hyschive commented:
@hfhsieh Thanks for the contribution! Before review, can you add a test problem (or modify `ParticleTest`) to validate this new functionality? For example, you can apply different analytical functions to different gas fields `F` and then compare the `(F, r)` plot between the values recorded by tracer particles and the analytical solutions. Here I have assumed that the analytical function is spherically symmetric but I'm sure you can have some better ideas. Make sure that the test is 3D with AMR to make it more comprehensive.

hyschive commented:
@ChunYen-Chen Could you help review it after @hfhsieh adds a unit test for this new feature? Thanks!

ChunYen-Chen commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,304 @@
+
+         for (int PID=PID0; PID<PID0+8; PID++)
+         {
+            if ( amr->patch[0][lv][PID]->NParType[(int)PTYPE_TRACER] > 0 )    GotYou = true;
+
+            if ( GotYou )  break;
```
How about combining them into:
``` C++
if ( amr->patch[0][lv][PID]->NParType[(int)PTYPE_TRACER] > 0 )
{
   GotYou = true;
   break;
}
```

hfhsieh replied:
Updated.

ChunYen-Chen reviewed (approved):
The PR looks great. I only have some minor comments.

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,304 @@
+         {
+            const long TVar = ParAttrIdx[v];
+
+//          2. prepare the mesh data for the patch group with particles (need NSIDE_26 for ParGhost>0 )
```
Remove the space before `)`. Also correct the one in `Par_UpdateTracerParticle()`

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,304 @@
+                  ParID = amr->patch[0][lv][PID]->ParList[p];
+
+                  if ( ParType[ParID] != PTYPE_TRACER )
+                     continue;
```
Done. Same for another one in line 281 (old number).

hfhsieh commented on src/Output/Output_DumpData_Total_HDF5.cpp:
```diff
@@ -1326,7 +1333,7 @@ void Output_DumpData_Total_HDF5( const char *FileName )
 //       output one particle attribute at one level in one rank at a time
 //       --> skip the last PAR_NATT_UNSTORED attributes since we do not want to store them on disk
-         for (int v=0; v<PAR_NATT_STORED; v++)
+         for (int v=0; v<PAR_NATT_STORED+PAR_NATT_MESH; v++)
```
Rename all `PAR_NATT_MESH` to `Par_NAtt_Mesh`.

ChunYen-Chen replied:
The comment is not updated yet.

hfhsieh replied:
My bad, I forgot the comment.

ChunYen-Chen reviewed (approved):
The updated test problem looks great. I only left some minor comments.

hfhsieh commented on src/Particle/Par_Init_Attribute_Mesh.cpp:
```diff
@@ -0,0 +1,129 @@
+      NItem = sscanf( Line, "%s", FirstItem );
+
+//    skip empty lines and lines starting with #
+      if ( NItem > 0  &&  FirstItem[0] != '#' )
```
Done.

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,183 @@
+
+               double EdgeL[3], EdgeR[3];
+
+               for (int d=0; d<3; d++) {
```
Done.

hyschive reviewed (commented):
@hfhsieh This PR looks great! I only have some minor comments. Once they are addressed, I would be happy to merge it right away.
- [x] Update `../doc/wiki` (this will be used to update the online wiki page directly; please consult @ChunYen-Chen if you need any help)
- [ ] Mention `OPT__OUTPUT_PAR_MESH` in the `ParticleTest/README`

hfhsieh commented on src/Auxiliary/Aux_Check_Parameter.cpp:
```diff
@@ -363,6 +363,11 @@ void Aux_Check_Parameter()
 #  endif
       Aux_Message( stderr, "WARNING : all output options are turned off --> no data will be output !!\n" );
 
+#  ifdef PARTICLE
+   if ( OPT__OUTPUT_PAR_MESH  &&  OPT__OUTPUT_TOTAL != OUTPUT_FORMAT_HDF5 )
+      Aux_Message( stderr, "WARNING : OPT__OUTPUT_PAR_MESH currently only supports OPT__OUTPUT_TOTAL=%d !!\n", OUTPUT_FORMAT_HDF5 );
+#  endif
```
Agreed. I also reset `OPT__OUTPUT_PAR_MESH = 0` if `TRACER` is disabled.

hfhsieh commented on src/Particle/Par_Init_Attribute_Mesh.cpp:
```diff
@@ -0,0 +1,128 @@
+
+   for (int v=0; v<Mesh_NAttr; v++)
+   {
+      amr->Par->Mesh_Attr      [v] = (real_par*) malloc( amr->Par->ParListSize*sizeof(real_par) );
```
Remove this line will cause segfault. Thus, I initialize `amr->Par->Mesh_Attr[v]` to `NULL` to avoid memory allocation.

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,184 @@
+
+// initialize the value to __FLT_MAX__ for all types of particles
+   for (int v=0; v<ParAttrNum;  v++)
+   for (int p=0; p<ParListSize; p++)
```
Corrected.

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,184 @@
+   {
+//    get the maximum number of particles in a single patch
+//    --> must use "NPar" instead of "NParType[(int)PTYPE_TRACER]" since currently
+//        both Vel_Temp[] and InterpParPos[] still allocate memory for non-tracer particles
```
Corrected

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,184 @@
+            if ( TVar & AllVar_PreparePatch )
+               Prepare_PatchData( lv, TimeNew, Var, NULL, ParGhost, 1, &PID0, TVar, _NONE,
+                                  OPT__FLU_INT_SCHEME, INT_NONE, UNIT_PATCH, NSIDE_26, IntPhase_No,
+                                  OPT__BC_FLU, BC_POT_NONE, MinDens_No, MinPres_No, MinTemp_No, MinEntr_No, DE_Consistency_No );
```
Updated

hfhsieh commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -0,0 +1,184 @@
+
+
+//          3. map quantities from mesh onto the tracer particles
+            for (int PID=PID0, P=0; PID<PID0+8; PID++, P++)
```
Updated

hfhsieh commented on example/test_problem/Hydro/ParticleTest/Input__Parameter:
```diff
@@ -188,6 +188,7 @@ OPT__OUTPUT_TOTAL             1           # output the simulation snapshot: (0=o
 OPT__OUTPUT_PART              0           # output a single line or slice: (0=off, 1=xy, 2=yz, 3=xz, 4=x, 5=y, 6=z, 7=diag) [0]
 OPT__OUTPUT_USER              0           # output the user-specified data -> edit "Output_User.cpp" [0]
 OPT__OUTPUT_PAR_MODE          0           # output the particle data: (0=off, 1=text-file, 2=C-binary) [0] ##PARTICLE ONLY##
+OPT__OUTPUT_PAR_MESH          0           # output the attributes of tracer particles mapped from mesh quantities -> edit "Input__Par_Mesh" [0] ##PARTICLE ONLY##
```
I am concerned that users may not realize that `OPT__OUTPUT_PAR_MESH` is enabled by default, and they might forget to include `Input__Par_Mesh`, which could lead to program crashes due to the missing file.

Also, could you remind me if there are any test problems that use tracer particles, so I can update them as well if we decide to enable this feature by default?

hyschive replied:
@hfhsieh 
- We should call `Aux_CheckFileExist()` in `Par_Init_Attribute_Mesh()` to verify the existence of the file `Input__Par_Mesh`. In that case, it should be fine to enable `OPT__OUTPUT_PAR_MESH` by default. Besides, this test problem belongs to a regression test instead of a scientific application, so I don't think it will affect other people's projects. What do you think?
- No other built-in test problem is using tracer particles.

hfhsieh replied:
@hyschive Agreed. `OPT__OUTPUT_PAR_MESH` is now enabled by default.

Additionally, I have updated `Par_Init_Attribute_Mesh()` to handle corner cases, where `Input__Par_Mesh` is missing or has no fields specified, to avoid segfault.

hfhsieh commented:
@hyschive I have updated the PR based on your suggestions, except for the one regarding enabling  `OPT__OUTPUT_PAR_MESH` by default.

I have also updated the `doc/wiki/Outputs.md` and the `README` for the `ParticleTest` problem. @ChunYen-Chen, Could you revise the changes in `Outputs.md` to ensure they are correct?

ChunYen-Chen commented on doc/wiki/Outputs.md:
```diff
@@ -105,6 +106,14 @@ This option is mainly for debugging purposes since the simulation snapshots
+Map mesh attributes to tracer particles and output them in the HDF5 snapshots.
+Edit `Input__Par_Mesh` to specify the desired mesh attributes. An example file can be found at `example/input/Input__Par_Mesh`.
+    * **Restriction:**
+Only applicable when enabling [TRACER](./Installation:-Simulation-Options.md#TRACER) and setting [OPT__OUTPUT_TOTAL](#OPT__OUTPUT_TOTAL) to 1.
```
The link of `TRACER` should be:
```markdown
[[TRACER | Installation:-Simulation-Options#TRACER]]
```

hfhsieh replied:
Corrected.

hyschive commented on src/Particle/Par_Output_TracerParticle_Mesh.cpp:
```diff
@@ -116,7 +120,7 @@ void Par_Output_TracerParticle_Mesh()
                Prepare_PatchData( lv, TimeNew, Var, NULL, ParGhost, 1, &PID0, TVar, _NONE,
                                   OPT__FLU_INT_SCHEME, INT_NONE, UNIT_PATCH, NSIDE_26, IntPhase_No,
-                                  OPT__BC_FLU, BC_POT_NONE, MinDens_No, MinPres_No, MinTemp_No, MinEntr_No, DE_Consistency_No );
+                                  OPT__BC_FLU, OPT__BC_POT, MinDens_No, MinPres_No, MinTemp_No, MinEntr_No, DE_Consistency_No );
```
[Bug] `OPT__BC_POT` is only defined when enabling `GRAVITY`.

hfhsieh replied:
This bug is fixed in the latest commit. 

However, when `GRAVITY` is disabled (and so for `MASSIVE_PARTICLES`), the linker will fail due to the functions `Prepare_PatchData_FreeParticleDensityArray()` and `Prepare_PatchData_InitParticleDensityArray()` are not defined, which are called in `Aux_ComputeProfile.cpp`, `Aux_FindExtrema.cpp`, and `Aux_FindWeightedAverageCenter.cpp`. Should I also fix it in this PR?

hyschive replied:
@hfhsieh Please file a separate PR to fix this issue. I think we should use `MASSIVE_PARTICLES` instead of `PARTICLE` in those files.

hyschive reviewed (changes requested):
@hfhsieh Minor comments added.

hfhsieh commented on src/Init/Init_Load_Parameter.cpp:
```diff
@@ -459,7 +459,7 @@ void Init_Load_Parameter()
 #  ifdef PARTICLE
    ReadPara->Add( "OPT__OUTPUT_PAR_MODE",       &OPT__OUTPUT_PAR_MODE,            0,               0,             2              );
-   ReadPara->Add( "OPT__OUTPUT_PAR_MESH",       &OPT__OUTPUT_PAR_MESH,            false,           Useless_bool,  Useless_bool   );
+   ReadPara->Add( "OPT__OUTPUT_PAR_MESH",       &OPT__OUTPUT_PAR_MESH,            true,            Useless_bool,  Useless_bool   );
```
Agreed. 

When `TRACER` is not defined, I force `OPT__OUTPUT_PAR_MESH` to be `false` and then we can remove the check and reset processes in `Aux_Check_Parameter.cpp` and `Init_ResetParameter.cpp`.

hfhsieh commented on src/Particle/Par_Init_Attribute_Mesh.cpp:
```diff
@@ -73,53 +73,59 @@ void Par_Init_Attribute_Mesh()
 
 // retrieve the number of mesh quantities specified in Input__Par_Mesh
-   const int Mesh_NAttr = Aux_CountRow( FileName );
+   const int Mesh_NAttr = ( Aux_CheckFileExist(FileName) ) ? Aux_CountRow( FileName ) : 0;
```
Agreed. 

Since it is already done in `Aux_CountRow()`, I reset this line to the previous version. Also, I add the `return` statement if `Mesh_NAttr = 0` such that we do not need check it again in the following code.

hyschive commented:
@hfhsieh Everything looks good. Thanks for the contribution!

Next, I'll review #348 and #351 to fully support tracer particles.
